workers/unity/Packages/io.improbable.gdk.tools/GenerateCode.cs

```diff
@@ -224,26 +253,133 @@ private static string[] ConstructArgs(string projectPath, string schemaCompilerP
+                Directory.Delete(CodegenExeDirectory, recursive: true);
             }
 
-            Generate();
+            Directory.CreateDirectory(CodegenExeDirectory);
+
+            var codegenLib = Path.GetFullPath(Path.Combine(Common.GetThisPackagePath(), CodeGenLibFile));
+
+            // Fix up symlinking for Mac
+            if (Application.platform == RuntimePlatform.OSXEditor)
+            {
+                var packageAttributes = File.GetAttributes(Common.GetThisPackagePath());
+                if (packageAttributes.HasFlag(FileAttributes.ReparsePoint))
+                {
+                    var process = RedirectedProcess.Command("pwd")
+                        .WithArgs("-P")
+                        .InDirectory(Common.GetThisPackagePath())
+                        .RedirectOutputOptions(OutputRedirectBehaviour.None)
+                        .RunAsync(CancellationToken.None)
+                        .Result;
+
+                    var realPath = string.Join("\n", process.Stdout).Trim();
+
+                    codegenLib = Path.GetFullPath(Path.Combine(realPath, CodeGenLibFile));
+                }
+            }
+
+            var result = RedirectedProcess.Command(Common.DotNetBinary)
+                .WithArgs("new", "gdk-for-unity-codegen",
+                    "--code-gen-lib-path", $"\"{codegenLib}\"")
+                .InDirectory(CodegenExeDirectory)
+                .RedirectOutputOptions(OutputRedirectBehaviour.None)
+                .Run();
+
+            if (result.ExitCode != 0)
+            {
+                throw new Exception("Failed to run.");
+            }
+        }
+
+        private static void UpdateModules()
+        {
+            var csprojXml = XDocument.Load(CodegenExe);
+            var projectNode = csprojXml.Element("Project");
+            var codegenDirs = FindDirInPackages(CodegenDir).ToList();
+
+            var gdkItemGroups = projectNode
+                .Elements("ItemGroup")
+                .Where(ele => ele.Element("GdkPackageSource") != null)
+                .ToDictionary(ele => ele.Element("GdkPackageSource").Attribute("Include").Value, ele => ele);
+
+            foreach (var dir in codegenDirs)
+            {
+                if (!gdkItemGroups.TryGetValue(dir, out var itemGroup))
+                {
+                    itemGroup = new XElement("ItemGroup");
+                    projectNode.Add(itemGroup);
+                }
+
+                itemGroup.RemoveAll();
+
+                var idEle = new XElement("GdkPackageSource");
+                idEle.SetAttributeValue("Include", dir);
+                itemGroup.Add(idEle);
+
+                var sourceDir = Path.Combine(dir, "Source");
+                if (Directory.Exists(sourceDir))
+                {
+                    var ele = new XElement("Compile");
+                    ele.SetAttributeValue("Include", Path.Combine(sourceDir, "**"));
+                    itemGroup.Add(ele);
```

Comment:
that whole if statement (and the ones below it) could be factored out into a method. that might also makes this a bit easier to read :)

Reply:
Discussed offline - this may not help legibility since its only generalizable in two places (the `Partial` one is different) without going down an `params (string nodeName, string attributeName)[]` approach on the helper method which seems a little bit overkill